File: src/baekjoon/dfs_bfs/P2667_단지번호붙이기.java
package baekjoon.dfs_bfs;

import java.util.*;

public class P2667_단지번호붙이기 {
    static int dx[] = new int[]{1, 0, -1, 0};
    static int dy[] = new int[]{0, 1, 0, -1};
    static int house[][];
    static boolean visited[][];
    static int N;
    static int result = 0;
    static int[] houseCount;

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        N = Integer.valueOf(scanner.nextLine());
        house = new int[N + 1][N + 1];
        visited = new boolean[N + 1][N + 1];
        for (int i = 0; i < N; i++) {
            String line = scanner.nextLine();
            for (int j = 0; j < N; j++) {
                house[i][j] = Character.getNumericValue(line.charAt(j));
            }
        }
        // === 세팅 ===
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                if (house[i][j] != 0 && visited[i][j] == false) {
                    result++;
                    BFS(i, j);
                }
            }
        }
        houseCount = new int[result+1];
        Arrays.fill(houseCount, 0);
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                if (house[i][j] != 0) {
                    houseCount[house[i][j]]++;
                }
            }
        }
        Arrays.sort(houseCount, 1, result+1);
        System.out.println(result);
        for (int i = 1; i < result + 1; i++) {
            System.out.println(houseCount[i]);
        }
    }


    static void BFS(int x, int y) {
        Queue<int[]> queue = new LinkedList<>();
        queue.add(new int[]{x, y});
        visited[x][y] = true;
        house[x][y] = result;
        while (!queue.isEmpty()) {
            int[] poll = queue.poll();
            int now_x = poll[0];
            int now_y = poll[1];
            for (int i = 0; i < 4; i++) {
                if (now_x + dx[i] <= N && now_x + dx[i] >= 0 && now_y + dy[i] <= N && now_y + dy[i] >= 0) {
                    if (house[now_x + dx[i]][now_y + dy[i]] != 0 && !visited[now_x + dx[i]][now_y + dy[i]]) {
                        queue.add(new int[]{now_x + dx[i], now_y + dy[i]});
                        visited[now_x + dx[i]][now_y + dy[i]] = true;
                        house[now_x + dx[i]][now_y + dy[i]] = result;
                    }
                }
            }
        }
    }
}

File: src/baekjoon/dfs_bfs/P2667_단지번호붙이기.py
result = []

# 입력
N = int(input())
map = []
check = [[0]*N for _ in range(N)]
for i in range(N):
    line = input()
    mapLine = []
    for j in range(N):
        mapLine.append(int(line[j]))
    map.append(mapLine)

def BFS(i, j):
    temp = 0
    queue = []
    queue.append((i, j))
    check[i][j] = 1
    while(len(queue) != 0):
        dy, dx = queue[0]
        del(queue[0])
        temp+=1
        if (dy-1) >= 0 and check[dy-1][dx] == 0 and map[dy-1][dx] != 0:
            queue.append((dy-1, dx))
            check[dy - 1][dx] = 1
        if (dy+1) < N and check[dy+1][dx] == 0 and map[dy+1][dx] != 0:
            queue.append((dy+1, dx))
            check[dy + 1][dx] = 1
        if (dx-1) >= 0 and check[dy][dx-1] == 0 and map[dy][dx-1] != 0:
            queue.append((dy, dx-1))
            check[dy][dx - 1] = 1
        if (dx+1) < N and check[dy][dx+1] == 0 and map[dy][dx+1] != 0:
            queue.append((dy, dx+1))
            check[dy][dx + 1] = 1
    result.append(temp)

for i in range(N):
    for j in range(N):
        if check[i][j] == 0 and map[i][j] != 0:
            BFS(i,j)

print(len(result))
result.sort()
for i in range(len(result)):
    print(result[i])